### ISSUE TITLE
How to stop a loop and waiting gracefully for end of current spawned sub-processes

### ISSUE THREAD

**Issue by TomFreudenberg** [amphp/parallel#54]
Hi Niklas @kelunik 

one more but different question.

I can't figure out how to stop a loop and wait until the current active sub-processes are finished.

My code so far:

```php
        // define worker pool
        $pool = new DefaultPool($cnt_threads);

        // loop through files
        foreach ($files as $file) {
          // create the task and push to pool
          $task = new FileTask($file);
          $pool->enqueue($task);
        }

        // event loop for parallel tasks
        Loop::run(function () use (&$pool) {

          // What to do when a signal is received
          foreach (array(SIGINT, SIGHUP, SIGTERM) as $signal) {
            Loop::unreference(
              Loop::onSignal($signal,
                function () use ($signal) {
                  echo "\nABORT: caught SIGNAL [" . $signal . "]! Trying gracefully shutdown ...\n";
                }
              )
            );
          }

          return $pool->shutdown();
        });

```

If I press now _CTRL-C_ the message comes up each time pressing keys and the Loop waits until current tasks are finished and stops then. 

BUT

If I send `kill -2 process-id` from console, just the message comes up but the Loop runs until all tasks are done.

  

Wondering what to do to gracefully break the `Loop::run()`. Do you have implement a special handler so that _CTRL-C_ works?

  

If I add `Loop::stop();` in the `onSignal` function, the Loop immediately stops and the main-app ends. BUT all current sub-processes are still running and visible as processes (`ps -ef`).

  
  
Thanks for your help
Tom

**Comment by bwoebi** [comment]
`DefaultPool::kill()` provides a method to unconditionally and immediately kill all the worker processes.

Use that instead of shutdown() then.

**Comment by TomFreudenberg** [comment]
Hi Bob @bwoebi and Niklas @kelunik 

thanks for your response. Neither of your suggestions will work in my conditions - still spooky. I had already tested `$pool->kill();` and `$pool->shutdown();` but did again after your help.

Here are the results:

#### Conditions:

Ubuntu 16.04 LTS - PHP 7.0.30

  

#### shutdown()

When using `$pool->shutdown()` in SignalHandler, I have to also remove the `shutdown()` call in my loop. Otherwise I get this exception:

> PHP Fatal error:  Uncaught Amp\Parallel\Context\StatusError: The pool was shutdown in /.../vendor/amphp/parallel/lib/Worker/DefaultPool.php:153

When call was removed and pressing _CTRL-C_ the loop stops as expected when the running tasks were finished.

When sending signal `kill -2`, the message was shown but the application was not aborted.

When sending signal `kill` twice, same exception as above.

  

#### kill()

When using `$pool->kill()` in SignalHandler and pressing _CTRL-C_ the loop stops as expected when the running tasks were finished.

When sending signal `kill -2`, the message was shown and the application stops immediately. All current active tasks still exists.

From time to time (1 out of 10) I got also an exception:

> In ChannelledStream.php line 50:
>  Sending on the channel failed. Did the context die?
> In ResourceOutputStream.php line 159:
>  The stream is not writable

  

So whatever I use, I can't get some reliable results.

Any ideas?

Thanks for your help.
Tom

**Comment by TomFreudenberg** [comment]
Addition:

It looks to me as if there is some special handling when *PRESSING* CTRL-C than sending signal SIGINT but can't figure out.

All my other Symfony commands from that same application works as expected. For those I register a SignalHandler by myself and check an interrupt during `while` processing.

**Comment by kelunik** [comment]
If you have another signal handler, that might conflict? Not sure...

You need to unregister the signal handler or keep a flag that you already triggered the shutdown. 

Not sure why your kill -2 acts differently.

**Comment by TomFreudenberg** [comment]
@kelunik 

> If you have another signal handler, that might conflict? Not sure...

No, not in this Command - I thought that maybe AMP does some magic?

> You need to ... keep a flag that you already triggered the shutdown.

Wouldn't it be more usefull that `shutdown()` handles such a flag by its own already?

> Not sure why your kill -2 acts differently

Spooky - have no clue

**Comment by kelunik** [comment]
> Wouldn't it be more usefull that shutdown() handles such a flag by its own already?

It does have such a flag and explicitly throws, but we could probably ignore duplicate calls for that operation. Usually we do that to prevent logic errors.

> No, not in this Command - I thought that maybe AMP does some magic?

There shouldn't be any related magic.

**Comment by TomFreudenberg** [comment]
> but we could probably ignore duplicate calls for that operation. Usually we do that to prevent logic errors.

Maybe you can add an argument to throw exception or not when calling `shutdown()` like

``` php
function shutdown (ignoreDuplicateCalls = false) { ... }
```

So it will be still compatible to previous behavior.

**Comment by TomFreudenberg** [comment]
Hi Niklas @kelunik, hi Bob @bwoebi 

I am still getting crazy to setup a graceful shutdown.

So I built a small Command example at: https://github.com/TomFreudenberg/amphp-sample-signals

There I create just sleeper tasks to have some work.

I catch the signals SIGINT, SIGTERM, SIGHUP and would expect, that they get ignored as I do nothing in their signal callbacks.

But behavior is different on my machines.

1. Pressing "CTRL-C" will show message BUT also stop the app AFTER the active running tasks are finished ????

2. Sending signal "-2" to the app will work as expected, just the message is shown but the app will run UNTIL ALL tasks are finished.

You can use

> ps -ef | grep "php.*sample_signal" | grep -v "grep" | sed -e 's/\s\+/ /g' | cut -d ' ' -f 2 | xargs -i\{\} kill -2 {}

in another window to find the process and send the signal.

  

I tested this on:

Ubuntu 16.04 LTS with PHP 7.0.30

and

Ubuntu 18.04 LTS with their latest PHP 7.2

  

May you check what happened?

Thanks for your help
Tom

  

P.S.: There are no other handlers registered than from Loop/NativeDriver. If you comment out the Loop::onSignal() part, than the app immediately stops on pressing CTRL-C as well as sending "-2" signal

**Comment by kelunik** [comment]
There are two things here...

1. Ctrl + C should behave the same as `kill -2` I guess, but according to https://unix.stackexchange.com/a/45435/89902 the signal is sent to all processes of the process group, not just the parent process, leading to the child processes exiting.
2. Calling `$pool->shutdown()` inside `Loop::run()` should directly shutdown the workers and not schedule more jobs.

**Comment by TomFreudenberg** [comment]
Hi Niklas,

thanks for answering on Sunday!

> Calling $pool->shutdown() inside Loop::run() should directly shutdown the workers and not schedule more jobs.

I rewrote the Loop into:

``` php
        // event loop for parallel tasks
        Loop::run(function () use (&$pool) {

          foreach (array(SIGINT, SIGTERM, SIGHUP) as $signal) {
            Loop::unreference(
              Loop::onSignal($signal,
                function () use ($signal, &$pool) {
                  echo "\nCaught SIGNAL [" . $signal . "]! normally this job should now continue ...\n";
                  return $pool->shutdown(); 
                }
              )
            );
          }
        });
```

But this not change anything.

CTRL-C will stop after CURRENT

kill -2 will stop AFTER ALL jobs

I also tried

``` php
                  echo "\nCaught SIGNAL [" . $signal . "]! normally this job should now continue ...\n";
                  $pool->shutdown();
                  return null;
```

**Comment by kelunik** [comment]
Yes, `$pool->shutdown()` doesn't seem to work correctly.

**Comment by TomFreudenberg** [comment]
Thanks for checking that, Niklas.

**Comment by TomFreudenberg** [comment]
Hi Niklas,

I also checked again CTRL-C and kill -2 behavior. Seems to me that also something may happens in Loop area.

If you change the code just to:

``` php
        // event loop for parallel tasks
        Loop::run(function () use (&$pool) {

          return $pool->shutdown(); 

        });
```

Then pressing CTRL-C and sending kill -2 does show the same behavior. Immediately the tasks are aborted and the app stops.

Also the process (group) list is killed when just sending signal -2 to parent process.

The process group looks like that:

<pre>
demouser 21147 15214  1 01:14 pts/1    00:00:00 php ./sample-signal sample:signal


demouser 21151 21147  0 01:14 pts/1    00:00:00 sh -c { ('/usr/bin/php7.0' -dhtml_errors=0 -ddisplay_errors=0 -dlog_errors=1 '/tmp/amphp-sample-signals/vendor/amphp/parallel/lib/Context/Internal/process-runner.php' '/tmp/amphp-sample-signals/vendor/amphp/parallel/lib/Worker/Internal/worker-process.php' 'Amp\Parallel\Worker\BasicEnvironment') <&3 3<&- 3>/dev/null & } 3<&0;pid=$!; echo $pid >&3; wait $pid; RC=$?; echo $RC >&3; exit $RC

demouser 21152 21151  0 01:14 pts/1    00:00:00 amp-process


demouser 21153 21147  0 01:14 pts/1    00:00:00 sh -c { ('/usr/bin/php7.0' -dhtml_errors=0 -ddisplay_errors=0 -dlog_errors=1 '/tmp/amphp-sample-signals/vendor/amphp/parallel/lib/Context/Internal/process-runner.php' '/tmp/amphp-sample-signals/vendor/amphp/parallel/lib/Worker/Internal/worker-process.php' 'Amp\Parallel\Worker\BasicEnvironment') <&3 3<&- 3>/dev/null & } 3<&0;pid=$!; echo $pid >&3; wait $pid; RC=$?; echo $RC >&3; exit $RC

demouser 21156 21153  0 01:14 pts/1    00:00:00 amp-process


demouser 21154 21147  0 01:14 pts/1    00:00:00 sh -c { ('/usr/bin/php7.0' -dhtml_errors=0 -ddisplay_errors=0 -dlog_errors=1 '/tmp/amphp-sample-signals/vendor/amphp/parallel/lib/Context/Internal/process-runner.php' '/tmp/amphp-sample-signals/vendor/amphp/parallel/lib/Worker/Internal/worker-process.php' 'Amp\Parallel\Worker\BasicEnvironment') <&3 3<&- 3>/dev/null & } 3<&0;pid=$!; echo $pid >&3; wait $pid; RC=$?; echo $RC >&3; exit $RC

demouser 21155 21154  0 01:14 pts/1    00:00:00 amp-process
</pre>

When using `Loop:onSignal` it is not reliable anymore.

Maybe that is a combined error you will get, when debugging `shutdown()`

  

Let me know if I may support with additional information

Thanks
Tom

  

P.S.: This does not match whats happened in https://unix.stackexchange.com/a/45435/89902

**Comment by TomFreudenberg** [comment]
Hi Niklas @kelunik 

any news on that issue? Thanks for a small update if possible.

Cheers
Tom

**Comment by kelunik** [comment]
@TomFreudenberg Yes, there's indeed an update. I assumed that we only schedule tasks for idle workers and keep them in the queue otherwise, but that doesn't seem to be the case, which in turn explains all jobs being processed instead of a graceful shutdown happening after the currently running jobs.

**Comment by TomFreudenberg** [comment]
Well, same as I detected from looking at amphp codes. Will "you" change that behavior or isn't this a bug anymore but wished feature?

**Comment by kelunik** [comment]
I'm not sure on that, I've already pinged @trowski on IRC for more information.

**Comment by TomFreudenberg** [comment]
Thanks for your update, lets see what will be decided by you guys.

**Comment by kelunik** [comment]
@trowski Having to change a test is a clear sign that it's "enough" of a BC break IMO.

**Comment by TomFreudenberg** [comment]
Hi Aaron @trowski 

I will check the branch tonight and let u know.

Thanks for working on it
Tom

**Comment by TomFreudenberg** [comment]
Hi Aaron @trowski 

that change from our branch seems to work for me.

I used for the Loop:

``` php
        // event loop for parallel tasks
        Loop::run(function () use (&$pool) {

          foreach (array(SIGINT, SIGTERM, SIGHUP) as $signal) {
            Loop::unreference(
              Loop::onSignal($signal,
                function () use ($signal, &$pool) {
                  echo "\nCaught SIGNAL [" . $signal . "]! normally this job should now continue ...\n";
                  return $pool->shutdown(); 
                }
              )
            );
          }
        });
```

And this worked fine, on CTRL-C and sending the signals.

Only question:

Is it correct, that I do not need the call for `$pool->shutdown();` anymore elsewhere?

otherwise I got the error still:

<pre>
PHP Fatal error:  Uncaught Amp\Parallel\Context\StatusError: 
The pool was shutdown in amphp-sample-signals/vendor/amphp/parallel/lib/Worker/DefaultPool.php:153
</pre>

**Comment by TomFreudenberg** [comment]
Hi Aaron @trowski 

here I updated the test project (file at: https://github.com/TomFreudenberg/amphp-sample-signals/blob/master/src/SignalDemo/SignalCommand.php)

When coding like there, everything works fine on my site.

Only enhancement I would like to have in addition, that "you" can call `->shutdown()` more than once without exception. Otherwise everyone have to save a flag by themself on every app. (see also https://github.com/amphp/parallel/issues/54#issuecomment-392093079)

**Comment by TomFreudenberg** [comment]
Hi Aaron @trowski  and Niklas @kelunik 

do you have any update infos about current status for this "bug"?

Thanks in advance for your feedback
Tom

**Comment by TomFreudenberg** [comment]
Hi Aaron @trowski and Niklas @kelunik

once again, do you have any update infos about current status for this "bug"?

Thanks in advance for any feedback
Tom

**Comment by trowski** [comment]
Merged ca051a0aa6db47ab6dc8560a87c9f26bcff15606 into master. That should address this issue.

I also removed the check for multiple calls to `shutdown()` in 526916d13ca0c3487f0cbc7b4f4f65e8f8510230.

**Comment by TomFreudenberg** [comment]
Hi Aaron @trowski 

thanks for fixing those 2 issues.

Tom

P.S.: Would you like me to stay with the test repository or do you like to take it over?
